src/mgos_imu_icm20948.c

```diff
@@ -0,0 +1,482 @@
+  struct mgos_imu_icm20948_userdata *iud = (struct mgos_imu_icm20948_userdata *)imu_user_data;
+
+  if (!dev) {
+    return false;
+  }
+
+  // Only initialize the ICM20948 if gyro hasn't done so yet
+  if (!iud->accgyro_initialized) {
+    if (!mgos_imu_icm20948_accgyro_create(dev->i2c, dev->i2caddr, imu_user_data)) {
+      return false;
+    }
+    iud->accgyro_initialized = true;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+
+  // REG2_ACCEL_CONFIG: ACCEL_DLPFCFG=101(12Hz); ACCEL_FS_SEL=10(8g); ACCEL_FCHOICE=1(Enable accel DLPF);
+  // ACCEL_SMPLRT_DIV_1: ACCEL_SMPLRT_DIV=0000(MSB);
+  // ACCEL_SMPLRT_DIV_2: ACCEL_SMPLRT_DIV=00001010(LSB);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_CONFIG, 0x15);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_SMPLRT_DIV_1, 0x00);
+  mgos_i2c_write_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_SMPLRT_DIV_2, 0x0a);
+
+  dev->scale = 8.f / 32767.0f;
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_read(struct mgos_imu_acc *dev, void *imu_user_data) {
+  uint8_t data[6];
+
+  if (!dev) {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 0)) {
+    return false;
+  }
+  if (!mgos_i2c_read_reg_n(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG0_ACCEL_XOUT_H, 6, data)) {
+    return false;
+  }
+
+  dev->ax = (data[0] << 8) | (data[1]);
+  dev->ay = (data[2] << 8) | (data[3]);
+  dev->az = (data[4] << 8) | (data[5]);
+
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_get_scale(struct mgos_imu_acc *dev, void *imu_user_data, float *scale) {
+  uint8_t fs = 0;
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+  if (!mgos_i2c_getbits_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_CONFIG, 1, 2, &fs)) {
+    return false;
+  }
+
+  switch (fs) {
+    case 0: *scale = 2; break;
+    case 1: *scale = 4; break;
+    case 2: *scale = 8; break;
+    case 3: *scale = 16; break;
+  }
+
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_set_scale(struct mgos_imu_acc *dev, void *imu_user_data, float scale) {
+  uint8_t fs = 0;
+
+  if (scale <= 2) {
+    fs = 0;
+    scale = 2.f;
+  } else if (scale <= 4) {
+    fs = 1;
+    scale = 4.f;
+  } else if (scale <= 8) {
+    fs = 2;
+    scale = 8.f;
+  } else if (scale <= 16) {
+    fs = 3;
+    scale = 16.f;
+  } else {
+    return false;
+  }
+
+  if(!mgos_imu_icm20948_change_bank(dev->i2c, dev->i2caddr, imu_user_data, 2)) {
+    return false;
+  }
+  if (!mgos_i2c_setbits_reg_b(dev->i2c, dev->i2caddr, MGOS_ICM20948_REG2_ACCEL_CONFIG, 1, 2, fs)) {
+    return false;
+  }
+  dev->opts.scale = scale;
+  dev->scale = dev->opts.scale / 32767.0f;
+
+  return true;
+}
+
+bool mgos_imu_icm20948_acc_get_odr(struct mgos_imu_acc *dev, void *imu_user_data, float *odr) {
+  uint16_t div  = 0;
+
```

Comment:
Please ensure that *odr is a valid pointer before trying to write to it later (causing a crash).
if (!odr) return false;